### PR TITLE
Fix #21725 - Render price on category list with correct precision for current locale

### LINKED ISSUE #21725: Price Precision wrong on product listing pages
### Preconditions (*)

1. Issue reproducible on: Magento version 2.2.7
2. Oman Rial selected as currency

### Steps to reproduce (*)

1. Visit product listing page - see that the products have prices with precision 2. 
2. Visit product details page - see that the products have prices with precision 3.

### Expected result (*)

1. Listing page should have consistent price precision with product page.
Oman Rial should have precision 3.
2. PHP code in Magento, does not take into account Locale precision while the product price js on product details page does, so it works on product details page but not on listing page

### Actual result (*)

1.Listing page has wrong precision

### Additional Info

Prices in Persian Rial currency are currently displayed with precision 2, but should be displayed with precision 0. - The solution should fix the precision for all currencies.

### PR DESCRIPTION
**Pull request by Bartlomiejsz**
### Description (*)

This PR changes price rendering on category view to include correct precision for current locale and currency.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#21725

### Manual testing scenarios (*)

1. Magento 2.3.0
2. Default locale set should be for Oman (ar_OM it's not by default available in Magento, need to be added in \Magento\Framework\Locale\Config but didn't check all locales to verify where it also occurs), Oman Rial (OMR) selected as currency
3. Visit product listing page - see that the products have prices with precision 3.
4. Visit product details page - see that the products have prices with precision 3.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @Bartlomiejsz. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Review by aleron75 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @aleron75, thank you for the review.
[ENGCOM-4538](https://jira.corp.magento.com/browse/ENGCOM-4538) has been created to process this Pull Request

**Comment by aleron75** [comment]
Hello @Bartlomiejsz thanks for your contribution.

Once merged, don't forget you can easily port this PR (if needed) to other branches with the Porting Tool, [read more here](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#porting-tool).

**Comment by VasylShvorak** [comment]
Hi @Bartlomiejsz during testing, I found more places when price is rendered with incorrect precision. And subtotal price on cart page is rendered with two comma separators(issue is present on 2.3-develop and not caused by PR). So price inconsistency is still present after your fix.

**Manual testing scenario:**
- Create simple product with Price=123,46
- Set Stores->Configuration->General->General->Locale Options->Arabic (Saudi Arabia)
- Go Stores->Configuration->General->Currency Setup
- Set Base Currency=Omani Rial, Default Display Currency=Omani Rial,Allowed Currencies=Omani Rial
- Check cart page, minicart and checkout page
- ❌ Price precision is not rendered correctly
#21829Cart
#21829CartSubtotal
#21829Checkout

@sivaschenko Could you check if testing scenario is valid for PR changes?

**Comment by sivaschenko** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, here is your [new Magento instance](https://pr-21829.instances.magento-community.engineering).
Admin access: [https://pr-21829.instances.magento-community.engineering/admin](https://pr-21829.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by sivaschenko** [comment]
@VasylShvorak the scenario looks valid to me. I think the sum is displayed correctly during checkout (looks like two different commas used)

@Bartlomiejsz would you be able to fix the price precision in the mini cart, shopping cart page and emails for consistency?

**Comment by Bartlomiejsz** [comment]
@sivaschenko yes, I will work on that

**Comment by Bartlomiejsz** [comment]
@VasylShvorak @sivaschenko could you test this one again with newest changes?

File: app/code/Magento/Sales/Model/Order.php
Comment (sivaschenko) [inline app/code/Magento/Sales/Model/Order.php]: Please change visibility to private for backward compatibility purpose

File: lib/internal/Magento/Framework/Pricing/Render/Amount.php
Comment (sivaschenko) [inline lib/internal/Magento/Framework/Pricing/Render/Amount.php]: Please change visibility to private for backward compatibility purpose

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for updates @Bartlomiejsz A couple of minor changes required for backward compatibility: please see my review comments

File: app/code/Magento/Sales/Model/Order.php
Comment (Bartlomiejsz) [reply app/code/Magento/Sales/Model/Order.php]: @sivaschenko done

File: lib/internal/Magento/Framework/Pricing/Render/Amount.php
Comment (Bartlomiejsz) [reply lib/internal/Magento/Framework/Pricing/Render/Amount.php]: @sivaschenko done

File: lib/internal/Magento/Framework/Pricing/Render/Amount.php
Comment (sivaschenko) [inline lib/internal/Magento/Framework/Pricing/Render/Amount.php]: @Bartlomiejsz sorry, looks like I missed that in the previous review, newly added parameters should be at the end of the parameters list (please move it after `$data`)

**Review by sivaschenko — CHANGES_REQUESTED** [review]

File: lib/internal/Magento/Framework/Pricing/Render/Amount.php
Comment (Bartlomiejsz) [reply lib/internal/Magento/Framework/Pricing/Render/Amount.php]: @sivaschenko done

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-4538](https://jira.corp.magento.com/browse/ENGCOM-4538) has been created to process this Pull Request

**Comment by VasylShvorak** [comment]
✔️ QA passed

**Comment by sivaschenko** [comment]
@paliarush can you please review and approve the added optional constructor parameters in this pull request

**Comment by paliarush** [comment]
New constructor arguments in the following @api classes look good to me:
- `app/code/Magento/Tax/Block/Item/Price/Renderer.php`

Please see if it is possible to avoid adding new argument to this class:
- `app/code/Magento/Sales/Model/Order.php`

**Comment by Bartlomiejsz** [comment]
@paliarush to fix price precision rendered from `\Magento\Sales\Model\Order` this new argument is required, to get correct price settings for current locale. I don't think it is possible in any other way...

**Comment by sivaschenko** [comment]
@paliarush would you suggest creating a temporary private getter method to retrieve the `FormatInterface` using the object manager without adding this dependency to the constructor?

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by paliarush** [comment]
@sivaschenko we have recently modified the rules regarding new optional constructor arguments in API classes. Now it is considered a patch change for most of the classes.

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-4538](https://jira.corp.magento.com/browse/ENGCOM-4538) has been created to process this Pull Request
:eight_spoked_asterisk: @sivaschenko, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Review by sidolov — CHANGES_REQUESTED** [review]
Hi @Bartlomiejsz , we found an issue with provided changes: huge performance degradation. 
Here is the build logs: 
```
cache-load
Load Customer Private Data (Catalog Browsing By Customer): 14.5% (mainline = 83 | team = 95) -> DEGRADATION
Open Category (Catalog Browsing By Customer): 13.8% (mainline = 260 | team = 296) -> DEGRADATION
Configurable Product 1 View (Catalog Browsing By Customer): 42.7% (mainline = 206 | team = 294) -> DEGRADATION
Open Category (Catalog Browsing By Guest): 14.8% (mainline = 244 | team = 280) -> DEGRADATION
Configurable Product 1 View (Catalog Browsing By Guest): 48.9% (mainline = 180 | team = 268) -> DEGRADATION
Search (Quick Search): 71.3% (mainline = 101 | team = 173) -> DEGRADATION
Search (Quick Search With Filtration): 72% (mainline = 100 | team = 172) -> DEGRADATION
Filter by Attribute 1 (Quick Search With Filtration): 72% (mainline = 100 | team = 172) -> DEGRADATION
Filter by Attribute 2 (Quick Search With Filtration): 32% (mainline = 75 | team = 99) -> DEGRADATION
Search (Advanced Search): 75.8% (mainline = 95 | team = 167) -> DEGRADATION
Open Category (Add To Cart By Guest): 14.8% (mainline = 244 | team = 280) -> DEGRADATION
Load Cart Section 1 (Add To Cart By Guest): 29.4% (mainline = 68 | team = 88) -> DEGRADATION
Load Cart Section 2 (Add To Cart By Guest): 40% (mainline = 70 | team = 98) -> DEGRADATION
Configurable Product 1 View (Add To Cart By Guest): 47.1% (mainline = 187 | team = 275) -> DEGRADATION
Load Cart Section 1 (Add To Cart By Guest) - Another one: 47.4% (mainline = 76 | team = 112) -> DEGRADATION
Load Customer Private Data (Add to Wishlist): 14.5% (mainline = 83 | team = 95) -> DEGRADATION
Simple Product 2 Add To Wishlist (Add to Wishlist) - Another one: 15.6% (mainline = 77 | team = 89) -> DEGRADATION
Load Wishlist Section 2 (Add to Wishlist): 14.6% (mainline = 82 | team = 94) -> DEGRADATION
Simple Product 3 Add To Wishlist (Add to Wishlist) - Another one: 20.3% (mainline = 79 | team = 95) -> DEGRADATION
Load Wishlist Section 3 (Add to Wishlist): 18.4% (mainline = 87 | team = 103) -> DEGRADATION
Simple Product 4 Add To Wishlist (Add to Wishlist) - Another one: 28.9% (mainline = 83 | team = 107) -> DEGRADATION
Load Wishlist Section 4 (Add to Wishlist): 22.5% (mainline = 89 | team = 109) -> DEGRADATION
Simple Product 5 Add To Wishlist (Add to Wishlist) - Another one: 36.8% (mainline = 87 | team = 119) -> DEGRADATION
Load Wishlist Section 5 (Add to Wishlist): 22.5% (mainline = 89 | team = 109) -> DEGRADATION
Open Category (Compare Products): 14.6% (mainline = 246 | team = 282) -> DEGRADATION
Configurable Product 1 View (Compare Products): 48.9% (mainline = 180 | team = 268) -> DEGRADATION
Configurable Product 1 Comparison Add (Compare Products) - Another one: 48.9% (mainline = 180 | team = 268) -> DEGRADATION
Compare Products (Compare Products): 29.4% (mainline = 68 | team = 88) -> DEGRADATION
Open Category (Checkout By Guest): 14.8% (mainline = 244 | team = 280) -> DEGRADATION
Load Cart Section 1 (Checkout By Guest): 29.4% (mainline = 68 | team = 88) -> DEGRADATION
Load Cart Section 2 (Checkout By Guest): 40% (mainline = 70 | team = 98) -> DEGRADATION
Configurable Product 1 View (Checkout By Guest): 47.1% (mainline = 187 | team = 275) -> DEGRADATION
Load Cart Section 1 (Checkout By Guest) - Another one: 47.4% (mainline = 76 | team = 112) -> DEGRADATION
Checkout start (Checkout By Guest): 13% (mainline = 92 | team = 104) -> DEGRADATION
Load Customer Private Data (Checkout By Customer): 14.5% (mainline = 83 | team = 95) -> DEGRADATION
Open Category (Checkout By Customer): 14% (mainline = 258 | team = 294) -> DEGRADATION
Load Cart Section 1 (Checkout By Customer): 23.5% (mainline = 85 | team = 105) -> DEGRADATION
Load Cart Section 2 (Checkout By Customer): 31.1% (mainline = 90 | team = 118) -> DEGRADATION
Configurable Product 1 View (Checkout By Customer): 42.9% (mainline = 205 | team = 293) -> DEGRADATION
Load Cart Section 1 (Checkout By Customer) - Another one: 37.5% (mainline = 96 | team = 132) -> DEGRADATION
Checkout start (Checkout By Customer): 11% (mainline = 109 | team = 121) -> DEGRADATION
Load Customer Private Data (Product Review By Customer): 14.5% (mainline = 83 | team = 95) -> DEGRADATION
Load Customer Private Data (Add To Cart By Customer): 14.5% (mainline = 83 | team = 95) -> DEGRADATION
Open Category (Add To Cart By Customer): 14% (mainline = 258 | team = 294) -> DEGRADATION
Load Cart Section 1 (Add To Cart By Customer): 23.5% (mainline = 85 | team = 105) -> DEGRADATION
Load Cart Section 2 (Add To Cart By Customer): 31.1% (mainline = 90 | team = 118) -> DEGRADATION
Configurable Product 1 View (Add To Cart By Customer): 42.9% (mainline = 205 | team = 293) -> DEGRADATION
Load Cart Section 1 (Add To Cart By Customer) - Another one: 37.5% (mainline = 96 | team = 132) -> DEGRADATION
Open Cart (Add To Cart By Customer): 26.9% (mainline = 134 | team = 170) -> DEGRADATION
Check Cart is Empty (Add To Cart By Customer): 15.8% (mainline = 76 | team = 88) -> DEGRADATION
Load Customer Private Data (Account management): 14.5% (mainline = 83 | team = 95) -> DEGRADATION
Open Order (Admin Returns Management): 26.9% (mainline = 134 | team = 170) -> DEGRADATION
Invoice Start (Admin Returns Management) - Another one: 23.3% (mainline = 103 | team = 127) -> DEGRADATION
Invoice Submit (Admin Returns Management) - Another one: 30.5% (mainline = 118 | team = 154) -> DEGRADATION
Credit Memo Start (Admin Returns Management) - Another one: 19.8% (mainline = 101 | team = 121) -> DEGRADATION
Credit Memo Submit - Full Refund (Admin Returns Management) - Another one: 30% (mainline = 120 | team = 156) -> DEGRADATION
Add Products (Admin Create Order): 20.1% (mainline = 179 | team = 215) -> DEGRADATION
Save Order (Admin Create Order) - Another one: 25% (mainline = 128 | team = 160) -> DEGRADATION
Save Invoice (Admin Create Order) - Another one: 25.4% (mainline = 126 | team = 158) -> DEGRADATION
Save Shipment (Admin Create Order) - Another one: 24.6% (mainline = 130 | team = 162) -> DEGRADATION
Open Order (Admin Edit Order): 29.3% (mainline = 123 | team = 159) -> DEGRADATION

loaded-classes
Set Shipping Address On Cart (GraphQL Set Shipping Address On Cart): 6% (mainline = 1175 | team = 1245) -> DEGRADATION
Set Billing Address On Cart (GraphQL Set Billing Address On Cart): 5.8% (mainline = 1199 | team = 1269) -> DEGRADATION
Set Shipping Address On Cart (GraphQL Checkout By Guest): 5.6% (mainline = 1254 | team = 1324) -> DEGRADATION

```

**Comment by Bartlomiejsz** [comment]
Hi @sidolov, could you let me know what tools are you using for such comparison? So I could test solution on my own after some improvements?
Thank you

**Comment by Bartlomiejsz** [comment]
Hi @sidolov, I added caching precision in local variable which should improve (and based on my local tests does indeed) performance. Still didn't get info what exactly tests your team uses, so I can't compare that in exactly same env though.
Also I improved unit tests for that part.

**Review by VladimirZaets — CHANGES_REQUESTED** [review]
@Bartlomiejsz can you please resolve conflicts? We measure performance by our internal builds. When you resolve conflicts we will help you with it

**Comment by Bartlomiejsz** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run Functional Tests EE,Functional Tests CE,Functional Tests B2B

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
@Bartlomiejsz what was the reason to update the `Amount` implementation with instead of `PriceCurrency` class?

**Comment by Bartlomiejsz** [comment]
@sivaschenko modification of `PriceCurrency` class wouldn't work here, since it implements `PriceCurrencyInterface`, which contains hardcoded precision - `2`. That's why I modified `Amount` class which seems to be some kind of wrapper on `PriceCurrency` class.
But this is pretty old PR which I tried to rebuild few times, and modification of `Amount` is from my first approach.
Now when I think about it - it would be probably better to introduce new interface and deprecate `PriceCurrencyInterface`, and get rid of this hardcoded precision.
Please let me know what would be better here in your opinion.

**Comment by sivaschenko** [comment]
@Bartlomiejsz I would update the default parameter value for `PriceCurrencyInterface` (without depreciation and reintroducing the new interface). But, in this case, we probably should deliver this change to 2.5 line.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

File: app/code/Magento/Catalog/Model/Product/Type/Price.php
Comment (aleron75) [inline app/code/Magento/Catalog/Model/Product/Type/Price.php]: As done before to improve readability, the _return early_ pattern can also be applied here, returning `$price` at the end of the loop and avoiding the `else` statement.

**Comment by ihor-sviziev** [comment]
Moving to on hold as this PR waiting for opening 2.5 develop branch

**Comment by ihor-sviziev** [comment]
@Bartlomiejsz Branch 2.5-develop was already created. Could you re-target this PR to the 2.5-develop branch?

**Comment by Bartlomiejsz** [comment]
Yes, @ihor-sviziev I'll try to work on this PR in upcoming days

**Comment by Bartlomiejsz** [comment]
@sivaschenko didn't really try to close this PR, it got automatically closed when I force pushed changes re-created on 2.5-develop branch, I will create new one for it

**Comment by ihor-sviziev** [comment]
@Bartlomiejsz seems like you force pushed branch with no changes

**Pull request closed without merging**
